### ISSUE TITLE
Pitch bend can cause aliasing or loss of upper harmonics

### ISSUE THREAD

**Issue by Diomendius** [Yoshimi/yoshimi#94]
When the pitch bend range is large enough, pitch bending up can shift upper harmonics above the Nyquist limit, causing aliasing, while pitch bending down causes any harmonics that were above the Nyquist limit at the note's base frequency to be inaudible even when the pitch shift would bring them below the limit.

With my recent changes to legato and portamento, portamento now triggers these issues just the same as pitch bend, as portamento now behaves nearly identically to pitch bend. Incidentally, the old implementation of portamento also displayed these issues, but in reverse; downward portamento would cause aliasing if the distance was large enough, as portamento in this case was analogous to starting a new note at the target pitch with the pitch wheel shifted up to bring the effective pitch to that of the old note. Likewise, portamento from a low to a high note would immediately drop the higher harmonics, which contributed to the slightly glitchy transitions between notes when legato and portamento were enabled. While the new portamento is smoother, the disadvantage is that the aliasing or dropped harmonics persist even after the target pitch is reached.

This happens because currently pitch bend (and now portamento) never triggers recalculation of note base frequency, which:
* For ADDsynth, rebuilds oscillator wavetables, cutting harmonics above the Nyquist limit via FFT.
* For SUBsynth, recalculates filters, skipping those above the Nyquist limit and attenuating those close to it.
* For PADsynth, sets the currently used wavetable to the one whose base frequency is closest to the desired base frequency. PADsynth precalculates these tables to remove frequency content close to or above the Nyquist limit similarly to how ADDsynth does for individual oscillators.

I think the most sensible solution is to change the base frequency whenever the effective frequency is too far from the base frequency. This should limit the amount of expensive computation done while still solving the issue. Keep in mind that currently this computation is done for every single NoteOn (at least in poly/mono mode), so "expensive" is relative here.

This would require rewriting pitch bend and portamento to be base frequency agnostic, and keeping track of the note's original pitch (nearly the same thing as the pitch of the MIDI note, if not for microtonality etc.) separately. I can see the potential for confusing variable names already. The best idea I have is to keep `basefreq` as it is and use something like `midinotefreq` for the actual note frequency before pitch bending or frequency offsets are applied.

When a note changes base frequency, it should crossfade just as legato notes do, as two waveforms with different frequency content will always have "seams" if you try to splice them together, even if you match up the phase of the frequencies they have in common. Legato crossfading needs to be changed as a prerequisite, as currently only two notes can fade at a time.

This isn't the only case where aliasing can occur due to pitch shifting. In particular, frequency modulation behaves very strangely at high modulator gain. It doesn't alias quite as I'd expect it to, though. When I set up a sine wave modulated by another sine wave with very low fixed frequency (-8 octaves, 440Hz base) and high enough modulator gain, I expect to hear aliasing when the effective frequency passes significantly above the Nyquist limit, audible as the pitch "bouncing" out of the ultrasonic range and back again before resuming the sine pattern it's supposed to be following, but while I do notice this, what's much more noticeable is that the pitch seems to "bounce" off some extreme low-frequency limit, too.

This can't happen due to normal sampling aliasing, so it must be something in the way FM is programmed in Yoshimi. I checked ZynAddSubFX too, and it behaves the same. I'm pretty sure it's because FM can actually cause the effective frequency to become negative, so the wavetable actually gets read backwards. This is probably a bug, but I guess it's arguable. It's also arguable that aliasing is desirable for FM, but it causes differences in the output depending on the sample rate, so I'd say it's not desirable either.

I'm not sure how FM can be tweaked to prevent aliasing, seeing as in the extreme case it could cause the effective frequency to jump by several octaves every few samples. Thankfully, FM this extreme doesn't seem to be particularly useful, so fixing this is probably unnecessary.

**Comment by abrolag** [comment]
This does seem to cover a wide range if known issues. my main concern is whether it might change existing being used over the normal range.

Also, Note_On is rather a sledgehammer that we get away with (mostly) because it doesn't happen too frequently, so I wonder what effect would tall this recalculation have?

Having said that, something that's bugged me for years is the significant (and variable) amount of time is wasted with all the memory allocation and removal.

**Comment by Diomendius** [comment]
> This does seem to cover a wide range if known issues. my main concern is whether it might change existing being used over the normal range.

I'm having some difficulty making sense of this, sorry. I think you might have lost one too many words along the way.

It's mainly only ADDsynth that has a truly hefty note constructor, though SUBsynth does do some work it probably doesn't need to. PADsynth generally has the right idea, though it lacks the bells and whistles that cause ADDsynth to be so heavy on computation and allocation.

I don't think my suggestion would have much of a performance impact. PADsynth's default parameters for how many wavetables to generate are a good starting point for how much the pitch should be allowed to change before the base frequency is recalculated. By default it uses two tables per octave across four octaves (taking up `4 octaves * 2 per octave * (128 * 1024) 128k default sample size * 4 bytes per float = 4194304 bytes` or exactly 4MiB, in case you were wondering), so let's say a note is allowed to keep its existing wavetable until it shifts up or down by half an octave. This means that `computeNoteParameters()` only needs to run two extra times for every octave of pitch bend. This is the same amount of work as two extra notes played, except it actually isn't because we skip a lot, including a lot of allocations, by running `computeNoteParameters()` rather than the entire note constructor.

To be technically accurate to how PADsynth does it, the threshold would be a quarter of an octave up or down in pitch and the new base frequency calculated would be another quarter octave further up or down from the point of recalculation, but the same reasoning still applies.

This brings me to one of two other potential improvements: ADDsynth doesn't actually need to allocate or recalculate its wavetables for each played note at all. In fact, there isn't a good reason for any of the synths to be allocating or deallocating anything outside of init, but I digress. Even the naive solution of preallocating one set of tables for each of the 128 possible MIDI notes could be an improvement, but there is no need to be so barbaric. PADsynth already provides a good solution, and adapting its approach to the individual oscillator tables of ADDsynth shouldn't be too difficult.

Using PADsynth's defaults for ADDsynth, eight sets of oscillator tables would be preallocated at different base frequencies and shared among all note instances. This is quite a lot fewer than an entire set of oscillator tables being allocated and computed for every single note played, ever.

Using two tables per octave means that the maximum difference is a quarter of an octave, or about 318 cents. Shifting 18.5kHz up by a quarter of an octave gives ~22kHz, which is less than the 22.05kHz Nyquist limit at a 44.1kHz sample rate. So, with two tables per octave and always using the nearest one, if the oscillators used contain no frequencies above ~18.5kHz, no aliasing will ever occur even at the standard CD sample rate. Conversely, when the desired frequency is below the base frequency, the 18.5kHz cutoff becomes ~15.56kHz, so all frequencies below that are always preserved.

If we don't care about preventing aliasing completely, only preventing aliasing within the audible range, the cutoff can be higher without increasing the number of tables used. If a 20kHz cutoff is used, the lowest frequency alias possible is at ~20.32kHz, which is not considered audible. The lowest clipping point for the upper harmonics in this case is at ~16.82kHz, which is technically within the audible range, but is probably inaudible for most people.

With modern computers, there's little reason to limit the number of table sets to just 2 per octave, and 6 per octave is plenty, giving a cutoff point for no aliasing (not even inaudible aliasing) of ~20.8kHz, with the lowest potential clipping point for harmonics being ~19.63kHz.

I also have some ideas about removing the allocation and table scanning from `Part.cpp`, but that really belongs in its own issue.

**Comment by kramlie** [comment]
> FM can actually cause the effective frequency to become negative, so the wavetable actually gets read backwards. This is probably a bug, but I guess it's arguable.

I also thought so at first, but it's not a bug I think. I believe commercial synths also behave like this. It's somewhat evident from the math behind it: If the modulation becomes strong enough to go backwards, then you must go backwards, otherwise going forwards afterwards isn't going to bring you back to the same point in the wavetable, and the result is that the note will drift out of tune.

I know some hardware synths used exponential FM modulation instead, which would get extremely close to zero, but never go negative. However, these have severe problems with staying in tune, and definitely not something I would recommend for a digital synth.

> ADDsynth doesn't actually need to allocate or recalculate its wavetables for each played note at all.

It has to in some cases, when you have randomized harmonics amplitudes or phases. But I agree that we can avoid it in many cases. The question is whether the complexity makes it worth fixing for only some cases, when we know there are some cases where it can't be fixed.

All in all I think recalculating the table is not such a big deal. It's a fully bounded operation, pretty much exactly O(n * log2 n) time complexity, where n is oscillator size.

That being said, it has occurred to me that if we explicitly made the optimization that wavetables can be pregenerated if randomized elements are not involved, we could:

1. Merge ADsynth and PADsynth into one engine, since they are essentially doing the same thing.
2. Expanding more upon it, it would be possible to implement a true wavetable engine, where we have multiple different wavetables we can morph between. This would be unreasonably expensive without pregenerated tables.

That's just some ideas, probably quite far off, but they are possible targets that would be unlocked.

**Comment by abrolag** [comment]
> > This does seem to cover a wide range if known issues. my main concern is whether it might change existing being used over the normal range.
> 
> I'm having some difficulty making sense of this, sorry. I think you might have lost one too many words along the way.

Apologies! Trying to carry too many different ideas in my head at the same time :(

My concern is the risk of noticeably changing existing voice patches and user's stored projects. I've been frustrated by this in the past with supposedly improved versions of both hardware and software synths.
I also inadvertently fell foul of it myself when bringing randomisation into our code(rather than a dependency). Fortunately it was spotted by Hermann who is more sensitive to such changes than I am.

Having said that, I'm generally in favour of anything that can improve performance *especially* if it reduces the note-on work. Currently, when running close to the wire, Xruns always occur at this point.

**Comment by Diomendius** [comment]
@kramlie
> > FM can actually cause the effective frequency to become negative, so the wavetable actually gets read backwards. This is probably a bug, but I guess it's arguable.
>
> I also thought so at first, but it's not a bug I think. I believe commercial synths also behave like this. It's somewhat evident from the math behind it: If the modulation becomes strong enough to go backwards, then you must go backwards, otherwise going forwards afterwards isn't going to bring you back to the same point in the wavetable, and the result is that the note will drift out of tune.
>
> I know some hardware synths used exponential FM modulation instead, which would get extremely close to zero, but never go negative. However, these have severe problems with staying in tune, and definitely not something I would recommend for a digital synth.

I would have thought that exponential (or logarithmic, if you want to look at it from the opposite direction) frequency modulation would be the most sensible interpretation; you're modulating perceived pitch rather than linear frequency. For most useful cases linear modulation is fine though, so it makes sense not to waste resources on something that provides little benefit. That said, you can precalculate 2^x or whatever for every sample in the modulator and store that in the table you read from, then use multiplication instead of addition.

I don't understand how the note could ever drift out of tune, though. A zero in the modulator always results in the carrier being output at its base frequency, it doesn't matter whether you let negative frequencies get used or clip the effective frequency at zero or whatever.

> > ADDsynth doesn't actually need to allocate or recalculate its wavetables for each played note at all.
> 
> It has to in some cases, when you have randomized harmonics amplitudes or phases. But I agree that we can avoid it in many cases. The question is whether the complexity makes it worth fixing for only some cases, when we know there are some cases where it can't be fixed.

You're right, I'd forgotten about that. We could be clever and store the individual harmonics in their own separate tables, so reconstructing a table with random harmonic amplitude and phase only involves computing the weighted sum of each harmonic starting from the appropriate offset, rather than having to recalculate everything and run it through an FFT and back again, but this already sounds fairly overengineered to me. Also, due to cache performance, excessive memoization like this might not even improve performance as much as you'd think, or at all.

The allocations can still be avoided, though. There will never be more than `POLIPHONY` sets of wavetables alive at once, so they can be stored in an array with the same layout as `partnote`. This is probably also overengineering, though.

Regardless of all this bikeshedding, my original suggestion should work fine. Hopefully I should be able to implement it sometime, now that I've done so much talking about it.

@abrolag 
> My concern is the risk of noticeably changing existing voice patches and user's stored projects. I've been frustrated by this in the past with supposedly improved versions of both hardware and software synths.

Backwards compatibility is always frustrating to deal with, whether providing it from the programming side or experiencing a lack of it on the end-user side. My original suggestion shouldn't change the sound of existing patches, it should only make the sound of pitch shifted notes the same as that of notes initially played at the same frequency. If, on the other hand, someone was relying on pitch bend causing aliasing as an effect, I'm not sure what to make of it, other than that it sounds a bit like [spacebar heating](https://xkcd.com/1172/).

There would actually be some other subtle effects, such as frequency scaling and the like applying (with a degree of granularity, admittedly) to pitch-bent notes rather than only caring about the base frequency. All of these effects should still be consistent with the way the note would sound if, say, legato was used instead of the pitch wheel.

My other suggestion of precomputing the oscillators shouldn't cause any audible changes either, but this depends much more on the implementation, as I'm sure there are a lot of edge cases, like making sure frequency scaling and the like are calculated based on the "real" frequency, not the oscillator base frequency.

The ideas about improving `Part.cpp` I mentioned before should improve NoteOn performance. I should probably profile Yoshimi before making significant claims, but there are quite a few big loops scanning through `partnote` for either empty slots or full slots that either don't need to be loops, or can be made to skip searching through empty slots. I'll try not to go into detail here, but I'll mention that the [find-first-set](https://en.wikipedia.org/wiki/Find_first_set) family of opcodes would be very useful, and GCC and Clang both provide `__builtin_ctz()`, which should compile even on targets which don't have applicable native instructions, though x86 always provides at least the `bsf` instruction.

**Comment by dbtx** [comment]
I'm that disposable idiot who needs or wants spacebar heating. If I want to imitate the sweeping ambient sounds in ([this track](https://www.youtube.com/watch?v=8G4sAVBvSYQ), I sidestep the note-on AA by using the pitch bend or envelopes (because that's what works), and I produce lots of aliasing, artistically. How else do you evenly fold all frequencies from negative infinity to infinity into the useful range, or simulate that?

It would be nice, but it would probably need to get much more complex in a bad way, to have user-toggled AA switches. Then such a "workaround" would never be "required" and then at least then the other question of whether existing instruments are rendered differently is moot. I don't have an existing instrument to point at or worry about preserving.

But then there are more nasty questions like "How many switches will there need to be? does a switch control AA across a part, or a kit item, or a single voice, or FM osc, or what?" and "does it default to being on or off?" And now I'm arguably filibustering again but the point is, sometimes the wrong thing is what you want, and you're allowed to want that, and I'm a hypocrite because I broke LFO already when I assumed that everyone should always want it to track the knobs in realtime. They might not.

**Comment by Diomendius** [comment]
You could use a higher sample rate so Yoshimi preserves the ultrasonic components, then pass the output through a bitcrusher to get the same effect. This also lets you adjust the Nyquist limit dynamically, so you can bring aliases further down into the audible range by either increasing the pitch of the synth, or decreasing the sampling rate.

That said, it would be nice to be able to control antialiasing in Yoshimi directly, even if the end result changes depending on the host sample rate. It shouldn't be too difficult to have a checkbox to disable oscillator antialiasing. Having the checkbox be per-oscillator would probably be appropriate, as the antialiasing in question happens in an FFT each time `OscilSmp` gets generated. I think antialiasing should be on by default, not just because that's the current behaviour, but because aliasing is an artifact of the representation of sound, not an inherent acoustic phenomenon.

In my opinion, it would be preferable to have some kind of sample rate agnostic control over aliasing, but the only way I can think of that wouldn't fail for a bunch of corner cases is to allow the internal sample rate of Yoshimi to be arbitrary, perhaps even on a per-voice basis. I wonder if there's an LV2 plugin somewhere that lets you host another LV2 plugin at an arbitrary sample rate, resampling the output before passing it to its own host?

**Comment by Ichthyostega** [comment]
Hi all,

just sent a general message to the mailinglist. We have technical questions here, but we also touch the concern of backward compatibility. From what I can see from this discussion here, the proposed change perfectly makes sense. Moreover, if done properly, we can ensure that the steady-state sound of the notes is not affected. And likewise, we can ensure that the performance is not degraded for anyone _not using portamento or pitch-bend at all._

However, my judgement from gut feeling is: the changed implementation will sound differently, at least in some area. We are doing a cross-fade. Sometimes, cross-fades can be arranged to be inaudible, yet given the right circumstances, crossfades can be very audible and add a tangible texture, especially to moving and changing sounds. And even if the crossfade itself is inaudible, we will remove audible alias frequencies, thus making the sound softer and less spicy.

So my personal stance is, in a nutshell:
 
- this is a very much valid proposal and could improve Yoshimi.
- but we can not just sneak it in. At least we need some kind of feature toggle.

**Comment by Ichthyostega** [comment]
To comment on the concern regarding allocations, I'd kindly ask you to take a step back and to try not to stick to the surface of matters. It might be compelling just to re-use some wavetable -- however, experience shows that we tend to overestimate the benefit of such local optimisations.

Not an allocation as such is _evil_ -- rather it is problematic when we use a general-purpose allocator (like the default heap allocator) for a very specific allocation trend. Rather, we should consider to write our own special-purpose allocator, which can exploit some additional structural knowledge that we have from our very controlled usage. Most notably: we know that our allocations can be pooled into blocks of identical size.

Will, Christian and myself had some in-depth discussion on that topic half a year ago, an IMHO the most important "red flag" we all saw are the ramifications for moving towards more concurrent calculations. Aggressively optimising allocations might hamper our abilities to compute more in parallel. Both topics should be approached together.

**Comment by abrolag** [comment]
Just a heads up. I've committed V 1.7.1 rc1.
There's a lot already there (currently undocumented), and it looks like there's going to be a lot more, so I'll accept bugfixes only for now.

**Comment by Diomendius** [comment]
As I found when trying to improve legato, crossfading is not an all-purpose solution. But to clarify, the desired effect in this case is for the individual frequency components to fade, as they are the only necessary point of difference between the tables.

Notably, harmonic amplitude/phase randomization shouldn't happen when recomputing the oscillators, for the same reason I prevented the new portamento from doing this. This actually prevents just calling `computeNoteParameters()`, as that function currently unconditionally re-randomizes harmonics.

Without harmonic randomization, a crossfade should be able to match the position in one table with its equivalent in the other, so the fade should be equivalent to fading only the frequency components which differ between the two. All else being equal, this should lead to a very subtle transition.

It may be desirable for basefreq-sensitive filters to readjust to the new effective frequency when pitch bend or portamento is in effect. This may even want to be at a separate threshold to the one used for recomputing oscillators. It's not obvious whether this is the "correct" behaviour, though, but it should be easy not to implement it and only recompute the oscillators; this would stay closest to the current behaviour.

As for allocations, I don't think it would be worthwhile to reuse oscillator tables because of the randomized harmonics feature. Making the lack of harmonic randomization a special case would add unnecessary complexity without improving performance when randomization is enabled.

However, most allocations at note construction are actually unnecessary given the current constraints Yoshimi imposes on itself; the number of objects needed while the synth is running is bounded entirely by `POLIPHONY` and a subset of the current patch configuration.

The memory required for a single note is fixed while the patch remains the same, and is probably still bounded by a known constant. Most allocations within the note objects themselves are either objects which either exist or not (mainly filters, envelopes and LFOs) or subvoices. Subvoices are potentially tricky, but should still be bounded, and there's no need to try to remove all the unnecessary allocations at once.

Placement-new can let us keep the existing constructor/destructor code as it is in most cases. If we were to bump the minimum language standard to C++17 we could also make use of `std::optional`, but I don't think it's necessary. Envelopes and LFOs don't allocate anything themselves, so placement-new can construct a new one over the old one without issue, otherwise allowing their reconfiguration outside the constructor isn't difficult. Filters are currently polymorphic (or rather, their `filter` member is polymorphic), which technically does not necessitate an allocation, but avoiding one is *extremely* hairy, not to mention ugly.

But the point is that there are a lot of allocations done which can be stored inline, which shouldn't have any implications for concurrent access, and would improve cache locality. The objects would still have the same lifetimes and the same number of distinct objects would still exist with distinct memory locations.

I have no idea how much of an impact reducing allocations would have, of course, so profiling Yoshimi is a must before throwing much effort at this.

**Comment by kramlie** [comment]
> Notably, harmonic amplitude/phase randomization shouldn't happen when recomputing the oscillators, for the same reason I prevented the new portamento from doing this. This actually prevents just calling `computeNoteParameters()`, as that function currently unconditionally re-randomizes harmonics.

Note that the code takes this into account by [reusing the same randomness seed](https://github.com/Yoshimi/yoshimi/blob/4e87c4f3297c7fc01c0ab7a1e3523ca63f2a5037/src/Synth/ADnote.cpp#L1135). So although it does re-randomize, the result is the same. Therefore calling it is perfectly fine, although it should still be avoided for performance reasons, if possible.

**Comment by Diomendius** [comment]
`OscilSmp`, which is an `OscilGen`, doesn't actually use `paramRNG` at all, though.

I actually need to correct myself here. `computeNoteParameters()` does not re-randomize oscillator harmonics at all, but the specific phase/amplitude values chosen for each harmonic depend on the base frequency the oscillator is generated at for some reason. I discovered this when I was working on legato and found that the timbre of the note would change when transitioning to a new note, but when transitioning back to the same note it would always have the same timbre that note originally had.

I explicitly reseed the oscillator in `legatoFadeIn()` to fix this:
https://github.com/Yoshimi/yoshimi/blob/4e87c4f3297c7fc01c0ab7a1e3523ca63f2a5037/src/Synth/ADnote.cpp#L691-L699

This actually makes fixing this issue slightly more complicated.

**Comment by kramlie** [comment]
Indeed, you're right. But `harmonicPrng`, used in `OscilGen`, functions the same way, so the general idea still applies. It is reinitialized by reusing the stored `randseed` value, which is only updated when calling `newrandseed`, as you pointed out.

And yes, this makes it considerably more complicated to implement pre-stored wavetables. I have not found any other solution to it than simply having two ways to do it: One with pregenerated tables, and one without, where the latter is triggered when you have a random element, like the harmonic amplitudes.

But we *kind of* have this already. The engine for pregenerated tables is just called PADsynth, and uses a much bigger table. But it isn't fundamentally different. This is why I'm contemplating whether it makes sense to merge the two. This would give some very nice benefits, for example, one could start using modulators on PADsynth sounds. The UI might be messy though...

**Comment by abrolag** [comment]
Adding modulators etc to PadSynth would damage it's greatest claim to fame. Very low running overhead - especially at note-on.

**Comment by Ichthyostega** [comment]
What Kristian proposes would mean an extensive rework of Yoshimi; however, the Idea sounds compelling.... there is a considerable amount of code duplicated between the two engines

**Comment by kramlie** [comment]
> Adding modulators etc to PadSynth would damage it's greatest claim to fame. Very low running overhead - especially at note-on.

Modulators do not generate any overhead if they are not used; this is true for ADsynth already, and would be true for either engine if implemented. It's much more likely that the low overhead is due to the pregenerated tables, which is exactly what we want to improve in ADsynth.

**Comment by abrolag** [comment]
Coming back to this. With V 1.7.1 out there and established in all distros (except debian) I think we can say it's open season on developments again. I'm certainly keen to see what people can come up with.
Apart from odd tweaks here and there (and possibly improvement of aftertouch) , I'm not thinking of doing anything much myself for now.

**Comment by abrolag** [comment]
Something just occurred to me.
Only since V 1.7.0 we've had realtime control of many of the waveform controls - including the the harmonics sliders. Would we have to lose these again?

**Comment by kramlie** [comment]
Yes, that is fundamentally the same problem as with the random harmonics; we can generate one waveform for each note in realtime, but we can't generated a whole set of waveforms for the entire key range. There has to be some sort of compromise to allow for pregenerated tables in some cases, but in other cases not. I have some ideas, but it's not something I've thought through all the way, so no use mentioning it until I have (or someone else has). Whoever attempts this needs to map this out in great detail before starting, I think.

**Comment by abrolag** [comment]
This is a repeat of a message I sent to several devs a while ago, only I don't have email addresses for all contributors. 

Just a heads up that I'll be starting to get Version 1.7.2 together in the week from 10th August, so if you've anything you'd like to add, please have it ready by then. If there is something really ground-breaking that you can't complete by that date let me know. It isn't fixed in stone!
